tracer/src/Datadog.Trace/Agent/AgentWriter.cs

```diff
@@ -402,15 +403,32 @@ SpanBuffer SwapBuffers()
                 spans = _statsAggregator?.ProcessTrace(spans) ?? spans;
                 bool shouldSendTrace = _statsAggregator?.ShouldKeepTrace(spans) ?? true;
                 _statsAggregator?.AddRange(spans);
+                var singleSpanSamplingSpans = new List<Span>(); // TODO maybe we can store this from above?
+
+                for (var i = spans.Offset; i < spans.Count; i++)
+                {
+                    if (spans.Array![i].GetMetric(Metrics.SingleSpanSampling.SamplingMechanism) is not null)
+                    {
+                        singleSpanSamplingSpans.Add(spans.Array![i]);
+                    }
+                }
 
                 // If stats computation determined that we can drop the P0 Trace,
                 // skip all other processing
-                if (!shouldSendTrace)
+                if (!shouldSendTrace && singleSpanSamplingSpans.Count == 0)
                 {
                     Interlocked.Increment(ref _droppedP0Traces);
                     Interlocked.Add(ref _droppedP0Spans, spans.Count);
                     return;
                 }
+                else if (!shouldSendTrace && singleSpanSamplingSpans.Count > 0)
+                {
+                    // change priority to be manual keep of the underlying trace context to notify agent
+                    spans.Array![spans.Offset].Context.TraceContext.SetSamplingPriority(SamplingPriorityValues.UserKeep);
+                    // note that we aren't incrementing _droppedP0Traces as we aren't dropping the trace entirely
+                    Interlocked.Add(ref _droppedP0Spans, spans.Count - singleSpanSamplingSpans.Count);
```

Comment:
This implementation looks good, but I'm double checking with other teams to make sure the header calculation is correct since it wasn't super clear to me from reading the RFC

Reply:
I think from the conversation this is good?

Reply:
Yeah, the implementation here is reasonable. Let's keep the implementation as-is and if we receive feedback later that it needs to be changed, we can update it then. Resolving